rfc/2020/conformance-validation/mayankshah1607/rfc.md

```diff
@@ -0,0 +1,484 @@
+- Contribution Name: Conformance Validation
+- Implementation Owner: [@mayankshah1607](https://github.com/mayankshah1607)
+- Start Date: 2020-02-23 (24th Feb, 2020)
+- Target Date: 
+- RFC PR:
+- Linkerd Issue: https://github.com/linkerd/linkerd2/issues/1096
+- Reviewers: 
+
+# Summary
+
+[summary]: #summary
+This project proposes a new test suite that shall be used for conformance validation. The new test suite shall be used to validate non-trivial network communication (HTTP, gRPC, websocket) among stateless and stateful workloads in the Linkerd data plane. The correctness of the interaction between the Linkerd control plane and the Kubernetes API Server will also be tested. This shall be done by carrying out extensive e2e tests of Linkerd features using a sample distributed application (data plane).
+
+# Problem Statement (Step 1)
+
+[problem-statement]: #problem-statement
+Linkerd has an extensive check suite that validates a cluster is ready to install Linkerd and that the installation was successful. These checks are, unfortunately, static checks. Because of the wide number of ways a Kubernetes cluster can be configured, users want a way to validate their specific install for conformance over time. The proposed project tackles this problem by allowing users to deploy sample workloads to their cluster and carry out extensive e2e (conformance) tests. 
+
+# Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+## Interaction with other features
+The e2e conformance tests proposed would be carried out outside the Linkerd CI. These tests would make use of the `kubectl` and `linkerd` binaries to interact with the cluster and with various features Linkerd can perform. The testing process can be configured easily using a YAML definition file (more details explained under “implementation details”). Such an approach gives the users the flexibility in diagnosing the state of their cluster w.r.t. Linkerd. 
+
+We initially start off by modifying the existing emojivoto application to have feature flags to enable/disable features such as MySQL, Redis, gRPC, websockets, etc. This is important because emojivoto is heavily used in the getting started process.
+
+## Implementation Details
+The test suite and its dependencies shall be packaged as a Docker container. This container shall run as a job in the cluster. In the Dockerfile, we define that the following must happen :
```

Comment:
> Outside of Sonobuoy, is it necessary to run your test framework as a container?
Nope. I've revamped this section and clearly described how the test framework will work with and without Sonobuoy.

Reply:
I don't have strong opinions here, but I always liked having a `docker run <something>` when its a one-time thing. Cloning the repo, and then running the tests on my local env ain't that hard, but having a container where they can pass both `kubeConfig` and configuration, and letting it run the tests feels really awesome without effecting my local env. WDYT?

Reply:
@Pothulapati That's actually possible with the current approach. The Sonobuoy section talks about having a Dockerfile (which is a requirement for Sonobuoy). This Dockerfile can definitely be utilised to run the tests locally as a part of the default mode - simply by building the image and then running it by passing the `./run.sh` args (as `ENTRYPOINT` is set to `./run.sh`) as you mentioned.

I wasn't sure how much information should have been added because I didn't want this section to extend for too long. Nevertheless, I have now added info about under the section that introduces the Dockerfile. Do have a look and let me know if that's what you're looking for! :smile: